exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/metadata/Metadata.java

```diff
@@ -230,68 +254,102 @@ private static boolean ignoreReadingMetadata(MetadataContext metaContext, Path p
-        columnTypeInfoSet.putAll(subTableMetadata.columnTypeInfo);
+        if (columnTypeInfoSet.isEmpty()) {
+          columnTypeInfoSet.putAll(subTableColumnTypeInfo);
+        } else {
+          for (ColumnTypeMetadata_v4.Key key : subTableColumnTypeInfo.keySet()) {
+            ColumnTypeMetadata_v4 columnTypeMetadata_v4 = columnTypeInfoSet.get(key);
+            if (columnTypeMetadata_v4 == null) {
+              columnTypeMetadata_v4 = subTableColumnTypeInfo.get(key);
+            } else {
+              if (subTableColumnTypeInfo.get(key).totalNullCount < 0 || columnTypeMetadata_v4.totalNullCount < 0) {
+                columnTypeMetadata_v4.totalNullCount = NULL_COUNT_NOT_EXISTS;
+              } else {
+                columnTypeMetadata_v4.totalNullCount = columnTypeMetadata_v4.totalNullCount + subTableColumnTypeInfo.get(key).totalNullCount;
+              }
+            }
+            columnTypeInfoSet.put(key, columnTypeMetadata_v4);
+          }
+        }
+        dirTotalRowCount = dirTotalRowCount + subTableMetadata.getTotalRowCount();
       } else {
         childFiles.put(file, fs);
       }
     }
-    ParquetTableMetadata_v3 parquetTableMetadata = new ParquetTableMetadata_v3(SUPPORTED_VERSIONS.last().toString(),
-                                                                                DrillVersionInfo.getVersion());
+    Metadata_V4.MetadataSummary metadataSummary = new Metadata_V4.MetadataSummary(SUPPORTED_VERSIONS.last().toString(), DrillVersionInfo.getVersion());
+    ParquetTableMetadata_v4 parquetTableMetadata = new ParquetTableMetadata_v4(metadataSummary);
     if (childFiles.size() > 0) {
-      List<ParquetFileMetadata_v3 > childFilesMetadata = getParquetFileMetadata_v3(parquetTableMetadata, childFiles, allColumns, columnSet);
-      metaDataList.addAll(childFilesMetadata);
-      // Note that we do not need to merge the columnInfo at this point. The columnInfo is already added
-      // to the parquetTableMetadata.
+      List<ParquetFileAndRowCountMetadata> ChildFileAndRowCountMetadata = getParquetFileMetadata_v4(parquetTableMetadata, childFiles, allColumnsInteresting, columnSet);
+      for (ParquetFileAndRowCountMetadata parquetFileAndRowCountMetadata : ChildFileAndRowCountMetadata) {
+        metaDataList.add(parquetFileAndRowCountMetadata.getFileMetadata());
+        dirTotalRowCount = dirTotalRowCount + parquetFileAndRowCountMetadata.getFileRowCount();
+        Map<ColumnTypeMetadata_v4.Key, Long> totalNullCountMap = parquetFileAndRowCountMetadata.getTotalNullCountMap();
+        if (columnTypeInfoSet.isEmpty()) {
+          columnTypeInfoSet.putAll(parquetTableMetadata.getColumnTypeInfoMap());
+        }
+        for (ColumnTypeMetadata_v4.Key columnName: totalNullCountMap.keySet()) {
+          ColumnTypeMetadata_v4 columnTypeMetadata_v4 = columnTypeInfoSet.get(columnName);
+          if ( columnTypeMetadata_v4.totalNullCount < 0 || totalNullCountMap.get(columnName) < 0) {
```

Comment:
If the first part of the || is true, it could mean that this is the first time we are initializing the `totalNullCount` .. so is that even a valid check ?  If the second part was >= 0,  we should not be setting the `NULL_COUNT_NOT_EXISTS`.

Reply:
Added a comment to explain this.